### PR TITLE
feat: validate major version being inferior to Java 18

### PR DESCRIPTION
**Pull request by tpetillot**
**Motivation:**
We do not support higher version of java than 8

**Modification:**
Throw 'UnsupportedJavaMajorVersionException' when a candidate have an
invalid java major version.

**Ref:** CLOUD-1424

### CODE REVIEW

**Review by slandelle — CHANGES_REQUESTED** [review]
Overly complicated. There's one thing Java Exceptions are good at (and streams are bad at): interrupts.
You could just throw your UnsupportedJavaMajorVersionException after loading the ClassNode, without carrying the extra 4 bytes luggage on each AsmClass instance.

**Comment by tpetillot** [comment]
@slandelle this does not allow to report invalid classes to the user. (which may be useful, when coming from a library for example)

File: src/main/java/io/gatling/scanner/JavaMajorVersion.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.gatling.scanner;
+
+import io.gatling.internal.asm.tree.ClassNode;
+
+public enum JavaMajorVersion {
+  JAVA_18(62, "Java 18");
```
Comment (slandelle) [inline src/main/java/io/gatling/scanner/JavaMajorVersion.java:22]: We don't care about Java 18. We care about Java 17 which is our highest supported version.
You can compute the Java major version from the bytecode version: bytecode version - 44.

**Review by slandelle — CHANGES_REQUESTED** [review]

File: src/main/java/io/gatling/scanner/JavaMajorVersion.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  JavaMajorVersion(int majorVersion, String noun) {
+    this.majorVersion = majorVersion;
+    this.noun = noun;
+  }
+
+  private static final int MAJOR_VERSION_CLASS_NODE_MASK = 0x0000FFFF;
+
+  public static void checkNode(ClassNode classNode) throws UnsupportedJavaMajorVersionException {
+    if ((classNode.version & MAJOR_VERSION_CLASS_NODE_MASK)
+        >= JavaMajorVersion.JAVA_18.majorVersion) {
```
Comment (slandelle) [inline src/main/java/io/gatling/scanner/JavaMajorVersion.java:36]: Change it to > Java17

**Review by slandelle — CHANGES_REQUESTED** [review]

File: src/main/java/io/gatling/scanner/JavaMajorVersion.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+      throw new UnsupportedJavaMajorVersionException(classNode, JavaMajorVersion.JAVA_18);
```
Comment (slandelle) [inline src/main/java/io/gatling/scanner/JavaMajorVersion.java:37]: We want to pass our max supported Java version and the actual Java version used. We don't want to break this when Java 19 is out.

**Review by slandelle — CHANGES_REQUESTED** [review]

File: src/main/java/io/gatling/scanner/AsmClass.java [hunk 4]
```diff
@@ -25,7 +25,8 @@ public class AsmClass {
   public final String parentName;
```
Comment (Isammoc) [inline src/main/java/io/gatling/scanner/AsmClass.java:25]: In commit message:
We do support java version above 8... up to 17 (included)

File: src/main/java/io/gatling/scanner/AsmClass.java [hunk 5]
```diff
@@ -25,7 +25,8 @@ public class AsmClass {
...
   public final boolean concrete;
 
-  AsmClass(ClassNode classNode) {
+  AsmClass(ClassNode classNode) throws UnsupportedJavaMajorVersionException {
+    JavaMajorVersion.checkNode(classNode, JavaMajorVersion.JAVA_17);
```
Comment (Isammoc) [inline src/main/java/io/gatling/scanner/AsmClass.java:29]: **design**
I'm concerned about having hard coded value that will change over time.
When we will release the gatling injector supporting Java 18 (or 22 for that matter) we should not have to change plugins for that!

Therefore, the scanner should not throw anything, I guess, but return the highest value found in the jar.
The user (of this library) will set its own limit.

At least, the acceptable value(s) should be an argument

File: src/main/java/io/gatling/scanner/JavaMajorVersion.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+import io.gatling.internal.asm.tree.ClassNode;
+
+public enum JavaMajorVersion {
+  JAVA_17(61, "Java 17");
+
+  public final int majorVersion;
+  public final String noun;
+
+  JavaMajorVersion(int majorVersion, String noun) {
```
Comment (Isammoc) [inline src/main/java/io/gatling/scanner/JavaMajorVersion.java:27]: default visibility should be avoided when possible!
```suggestion
  private JavaMajorVersion(int majorVersion, String noun) {
```

File: src/main/java/io/gatling/scanner/AsmClass.java (hunk 5 above)
Comment (slandelle) [reply src/main/java/io/gatling/scanner/AsmClass.java:29]: It was my proposal to have this centralized in the scanner to not duplicate the error message.
But I've come to realize that the error messages must be specific to the caller. Typically, we want to provide a hint on how to configure the build tool to have `--release` set.

I like your proposal of having the scanner report the highest Java version used in the package and let the caller deal with it.

Comment (slandelle) [reply src/main/java/io/gatling/scanner/AsmClass.java:29]: Now, if we don't want to have to update the clients when changing the max supported Java version, it means we need an API to advertise it.

**Review by slandelle — CHANGES_REQUESTED** [review]

**Pull request closed without merging**